File: LeetCode/739.daily-temperatures.cpp
/*
 * @lc app=leetcode id=739 lang=cpp
 *
 * [739] Daily Temperatures
 */

// @lc code=start
class Solution
{
public:
    vector<int> dailyTemperatures(vector<int> &temperatures)
    {
        vector<int> out(temperatures.size());
        stack<int> s;
        for (int i = temperatures.size() - 1; i >= 0; --i)
        {
            while (!s.empty() && temperatures[i] >= temperatures[s.top()])
            {
                s.pop();
            }
            if (s.empty())
            {
                out[i] = 0;
            }
            else
            {
                out[i] = s.top() - i;
            }
            s.push(i);
        }
        return out;
    }
};
// @lc code=end

File: LeetCode/739.daily-temperatures.py
#
# @lc app=leetcode id=739 lang=python3
#
# [739] Daily Temperatures
#


# @lc code=start
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        answer = [0] * len(temperatures)

        for i, temperature in enumerate(temperatures):
            while stack and stack[-1][0] < temperature:
                _, k = stack.pop()
                answer[k] = i - k

            stack.append((temperature, i))

        return answer


# @lc code=end
